_posts/2021-08-18-Day-8-Post.md

```diff
@@ -0,0 +1,123 @@
+
+## Const referencing
+
+A `const` reference can be passed into a function to make an imutable variable. Once a const variable is passed into a function, it cannot be changed within the function. However it is important to maintian *const correctness* when use const references. A `const` reference can accept both `const` and non-`const` data, but non-`const` reference can only take non-`const` objects. Therefore, a non-`const` reference cannot be made of a `const` object, you can't try to refer and change a `const` object. It's generally recommending to make referene arguments const if you do not have any intention of modifying the data, and is importing for maintaining the integrity of large data structures, like vectors and strings. It is much safer to write functions that do not mutate their input parameters, and to avoid global parameters.
+```cpp
+#include <iostream>
+double convertFtoC(double const temperature)
+{
+    return (temperature - 32)*(5.0/9.0);
+}
+
+int main(void)
+{
+    double temperatureF = 212.0;
+    double tempuratureC=convertFtoC(temperatureF);
+    std::cout<<"Original temperature is " <<temperatureF<<"F."<<std::endl;
+    std::cout<<"New temperature is " <<temperatureC<<"C."<<std::endl;
+    return 0;
+}
+```
+
+## Function Overloading
+
+In C++, we can have multiple functions of the same name, so long as the functions have different signatures. This is called function overloading Because the compiler will determine which function is used based on the data types and numbers of the argument. It is important to note that functions cannot be overloaded by return type, meaning you cannot have two functions that input the same data type, but return two different types of data.  For examples, we can have two `convertFtoC` functions, where one takes a `double` and the toher takes a `std::vector<double>`
+
+```cpp
+#include <iostream>
+#include <vector>
+
+void convertFtoC(double & temperature)
+{
+    temperature= (temperature - 32)*(5.0/9.0);
+}
+
+void convertFtoC(std::vector<double> & temperatures)
+{
+    for (auto & t : temperatures)
+    {
+       convertFtoC(t);
+    }
+}
+```
+## Exceptions
+
+C++ gives us several tools to test for errors. Because C++ requires the code to be free of errors upon compilation, any errors in the code will force the program to terminate. Therefore it is usful to use exceptions to catch areas of code that may be prone to errors to facillitate identifying sources of bug.  The `try` statement allows us to test a section of code for errors. If an error is encountered in the `try` block, a `catch` statement can then be used to define a section of code to be executed. At any point a error is caught, a `throw` statement allows us to make custom errors before the program is aborted.
```

Comment:
```suggestion
C++ gives us several tools to test for errors. Because C++ requires the code to be free of errors upon compilation, any errors in the code will force the program to terminate. Therefore it is useful to use exceptions to catch areas of code that may be prone to errors to facilitate identifying sources of bug.  The `try` statement allows us to test a section of code for errors. If an error is encountered in the `try` block, a `catch` statement can then be used to define a section of code to be executed. At any point a error is caught, a `throw` statement allows us to make custom errors before the program is aborted.
```